### PR TITLE
Fix spectator not starting from current player position

### LINKED ISSUE #17267: Spectating mid-gameplay no longer seeks to first received frame time
Regressed in https://github.com/ppy/osu/pull/17026/commits/bb8caabb8be3e0391fee7adede8f0bed09948737 due to `MasterGameplayClockContainer` resetting current time on `LoadComplete` rather than BDL. 

`SpectatorPlayer` can receive the first frame and seek to it before `MasterGameplayClockContainer` reaches `LoadComplete` state, therefore, after https://github.com/ppy/osu/pull/17026/commits/bb8caabb8be3e0391fee7adede8f0bed09948737, `MasterGameplayClockContainer` would overwrite any seeks done before `LoadComplete`, and regress spectating mid-gameplay.

I'm unsure about a solution to this, perhaps the sanest would be to schedule the first frame seek in `SpectatorPlayer` to run after `Player` updates children (i.e. `ScheduleAfterChildren`)

### Discussed in https://github.com/ppy/osu/discussions/17266

<div type='discussions-op-text'>

<sup>Originally posted by **minasminas** March 16, 2022</sup>
The singleplayer spectate doesn't seem to work anymore. It did work a few weeks ago (don't know which update broke it). Multiplayer spectate is working good as usual by the way.

While testing this issue I also made sure that the user is really playing the map and isn't in the menus or pausing.

Here is a short video of the issue:

https://user-images.githubusercontent.com/81645139/158482669-6692fe95-93f8-4b8b-a445-d02811147468.mp4

Logs:
[performance.log](https://github.com/ppy/osu/files/8257473/performance.log)
[runtime.log](https://github.com/ppy/osu/files/8257475/runtime.log)
[updater.log](https://github.com/ppy/osu/files/8257476/updater.log)
[database.log](https://github.com/ppy/osu/files/8257477/database.log)
[network.log](https://github.com/ppy/osu/files/8257478/network.log)

</div>

### PR DESCRIPTION
**Pull request by peppy**
Closes https://github.com/ppy/osu/issues/17267.

Changes to `PlayerLoader` are only required for testing. Can be split out into a separate PR if preferred, but I didn't bother because they are file (and commit) level isolated.

I think reading through the changes and attempting to understand from code is best (if it isn't 100% straightforward what has changed and why then I think further refactor may be required) so I'm not going to overly explain this one.

### CODE REVIEW

**Comment by peppy** [comment]
Note that I haven't touched the fact that int the first frame after `LoadComplete`, `Player` is not-present. I honestly don't know if that should be fixed or how to fix it (there's probably other screens in a similar scenario) so ensured the solution offered here doesn't rely on that part anymore.

File: osu.Game/Screens/Play/MasterGameplayClockContainer.cs [hunk 1]
```diff
@@ -90,26 +89,33 @@ protected override void LoadComplete()
...
-                if (firstStoryboardEvent != null)
-                    startOffset = Math.Min(startOffset, firstStoryboardEvent.Value);
-
-                // some beatmaps specify a current lead-in time which should be used instead of the ruleset-provided value when available.
-                // this is not available as an option in the live editor but can still be applied via .osu editing.
-                if (beatmap.BeatmapInfo.AudioLeadIn > 0)
-                    startOffset = Math.Min(startOffset, firstHitObjectTime - beatmap.BeatmapInfo.AudioLeadIn);
-            }
+                // sane default provided by ruleset.
+                double offset = gameplayStartTime;
```
Comment (frenzibyte) [inline osu.Game/Screens/Play/MasterGameplayClockContainer.cs:95]: Would rename `gameplayStartTime` to not conflict with the introduced property `GameplayStartTime` (maybe `rulesetGameplayStartTime`).

**Review by frenzibyte — COMMENTED** [review]
Haven't deeply checked this, but the direction sounds good in general. 

`startAtGameplayStart` could potentially be removed in favour of `Player.SetGameplayStartTime`/`GameplayClockContainer.Reset(gameplayStartTime: ...)`.

File: osu.Game/Screens/Play/MasterGameplayClockContainer.cs (hunk 1 above)
Comment (peppy) [reply osu.Game/Screens/Play/MasterGameplayClockContainer.cs:95]: Wow, I didn't even see this variable conflict.

**Comment by peppy** [comment]
I've made quite serious changes as I misunderstood the constructor parameter initially (or missed it). Note that while this is still fine for review purposes, I'm still stacking one remaining test issue, which is that `TestSceneMultiplayerPlayer` will occasionally fail (both the ctor and the test method):

JetBrains Rider-EAP 2022-03-17 at 18 28 26

```csharp
TearDown : System.AggregateException : One or more errors occurred. (StartGameplay should not be called when the gameplay clock is already running)
  ----> System.InvalidOperationException : StartGameplay should not be called when the gameplay clock is already running
--TearDown
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at osu.Framework.Extensions.TaskExtensions.WaitSafely(Task task)
   at osu.Framework.Testing.TestScene.checkForErrors()
   at osu.Framework.Testing.TestScene.RunTestsFromNUnit()
--InvalidOperationException
   at osu.Game.Screens.Play.Player.StartGameplay() in /Users/dean/Projects/osu/osu.Game/Screens/Play/Player.cs:line 982
   at osu.Game.Screens.OnlinePlay.Multiplayer.MultiplayerPlayer.<onMatchStarted>b__22_0() in /Users/dean/Projects/osu/osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerPlayer.cs:line 179
   at osu.Framework.Threading.ScheduledDelegate.InvokeTask()
   at osu.Framework.Threading.ScheduledDelegate.RunTaskInternal()
   at osu.Framework.Threading.Scheduler.Update()
   at osu.Framework.Graphics.Drawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Graphics.Containers.CompositeDrawable.UpdateSubTree()
   at osu.Framework.Platform.GameHost.UpdateFrame()
   at osu.Framework.Platform.HeadlessGameHost.UpdateFrame()
   at osu.Framework.Threading.GameThread.processFrame()
--- End of stack trace from previous location ---
   at osu.Framework.Platform.GameHost.<>c__DisplayClass126_0.<abortExecutionFromException>b__0()
   at osu.Framework.Threading.ScheduledDelegate.InvokeTask()
   at osu.Framework.Threading.ScheduledDelegate.RunTaskInternal()
   at osu.Framework.Threading.Scheduler.Update()
   at osu.Framework.Threading.GameThread.processFrame()
   at osu.Framework.Threading.GameThread.RunSingleFrame()
   at osu.Framework.Platform.ThreadRunner.RunMainLoop()
   at osu.Framework.Platform.GameHost.windowUpdate()
   at osu.Framework.Platform.GameHost.Run(Game game)
   at osu.Framework.Testing.TestScene.<>c__DisplayClass44_0.<SetupGameHostForNUnit>b__0()
   at System.Threading.Tasks.Task.InnerInvoke()
   at System.Threading.Tasks.Task.<>c.<.cctor>b__272_0(Object obj)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)



-----

One or more child tests had errors
  Exception doesn't have a stacktrace



[performance] 2022-03-17 18:28:54 [verbose]: TextureAtlas initialised (1024x1024)
[runtime] 2022-03-17 18:28:54 [verbose]: 💨 Class: TestSceneMultiplayerPlayer
[runtime] 2022-03-17 18:28:54 [verbose]: 🔶 Test:  TestConstructor
[runtime] 2022-03-17 18:28:55 [verbose]: Opened realm "/var/folders/qt/mf5nyxrn07s0_mwpb_t8bl5h0000gn/T/of-test-headless/TestSceneMultiplayerPlayer-76193df2-95ab-493c-b558-e3eda2b8b72f/client.realm" at version 14
[runtime] 2022-03-17 18:28:55 [verbose]: 🔸 Step #2 exit all screens
[runtime] 2022-03-17 18:28:55 [verbose]: 🔸 Step #3 join room
[runtime] 2022-03-17 18:28:55 [verbose]: 🔸 Step #4 wait for room join
[runtime] 2022-03-17 18:28:55 [verbose]: ✔️ 14 repetitions
[runtime] 2022-03-17 18:28:55 [verbose]: 🔸 Step #5 set beatmap
[runtime] 2022-03-17 18:28:55 [verbose]: 🔸 Step #6 initialise gameplay
[runtime] 2022-03-17 18:28:55 [verbose]: ScreenTestScene screen changed → MultiplayerPlayer
[runtime] 2022-03-17 18:28:55 [verbose]: 📺 ScreenTestScene(OsuScreenStack)#164(depth:1) loading MultiplayerPlayer#283
[runtime] 2022-03-17 18:28:55 [verbose]: 🔸 Step #7 wait for player to be current
[performance] 2022-03-17 18:28:56 [verbose]: TextureAtlas size exceeded 1 time(s); generating new texture (1024x1024)
[runtime] 2022-03-17 18:28:56 [verbose]: 📺 ScreenTestScene(OsuScreenStack)#164(depth:1) entered MultiplayerPlayer#283
[runtime] 2022-03-17 18:28:56 [verbose]: 📺 BackgroundScreenStack#571(depth:1) loading BackgroundScreenBeatmap#603
[runtime] 2022-03-17 18:28:56 [verbose]: ✔️ 1383 repetitions
[runtime] 2022-03-17 18:28:56 [verbose]: 🔸 Step #8 start gameplay
[runtime] 2022-03-17 18:28:56 [verbose]: 📺 BackgroundScreenStack#571(depth:1) entered BackgroundScreenBeatmap#603
[runtime] 2022-03-17 18:28:56 [error]: An unhandled error has occurred.
[runtime] 2022-03-17 18:28:56 [error]: System.InvalidOperationException: StartGameplay should not be called when the gameplay clock is already running
[runtime] 2022-03-17 18:28:56 [error]: at osu.Game.Screens.Play.Player.StartGameplay() in /Users/dean/Projects/osu/osu.Game/Screens/Play/Player.cs:line 982
[runtime] 2022-03-17 18:28:56 [error]: at osu.Game.Screens.OnlinePlay.Multiplayer.MultiplayerPlayer.<onMatchStarted>b__22_0() in /Users/dean/Projects/osu/osu.Game/Screens/OnlinePlay/Multiplayer/MultiplayerPlayer.cs:line 179
[runtime] 2022-03-17 18:28:56 [error]: at osu.Framework.Threading.ScheduledDelegate.InvokeTask()
[runtime] 2022-03-17 18:28:56 [error]: at osu.Framework.Threading.ScheduledDelegate.RunTaskInternal()
[runtime] 2022-03-17 18:28:56 [error]: at osu.Framework.Threading.Scheduler.Update()
[runtime] 2022-03-17 18:28:56 [error]: at osu.Framework.Graphics.Drawable.UpdateSubTree()
```

If someone beats me to figuring out why I will not complain.

**Comment by peppy** [comment]
This is ready for review again ._.

File: osu.Game/Screens/Play/MasterGameplayClockContainer.cs [hunk 2]
```diff
@@ -49,33 +49,38 @@ public class MasterGameplayClockContainer : GameplayClockContainer
         private readonly BindableDouble pauseFreqAdjust = new BindableDouble(1);
```
Comment (frenzibyte) [inline osu.Game/Screens/Play/MasterGameplayClockContainer.cs:49]: It's both funny and scary how something like this can cause tests to fail in a brain-melting way (definitely worth adding test coverage for):
```suggestion
        private readonly BindableDouble pauseFreqAdjust = new BindableDouble();
```
(`GameplayClockContainer` is now paused by default, so the adjustment should be `0`, otherwise will be stuck to `1` until `IsPaused` is toggled off and back on)

To explain why this is only failing when multiple test cases run in the same session (and I'm gonna pick the `TestSeekPerformsInGameplayTime(0.5d, -200.0d, true)` test case as an example while explaining): Since `pauseFreqAdjust` is set to 1 by default (which is incorrect), the first `ProcessFrame` call that the offset clocks receive are when `Reset` is called in the test case:
https://github.com/ppy/osu/blob/7cb7b03cee678db55530d8134d39295089a6a216/osu.Game.Tests/Gameplay/TestSceneMasterGameplayClockContainer.cs#L86-L88

And, since another test has ran in the same session, `OsuSetting.AudioOffset` will already be populated with a certain offset (let's say -200ms) at the point of calling `Reset`, which ends up making `userGlobalOffsetClock.RateAdjustedOffset` matching that offset at the *current* rate (which is 1, not 0.5, since the track is recreated per-run, and the rate adjustment is applied in a later step).

Then, since on this test case the clock remains stopped, the second `ProcessFrame` call only happens during the `Seek(2500)` step, and the failure occurs there, since the following happens when `Seek` gets called:
 - The *current* `totalAppliedOffset` (which is going to be -200ms from above) is subtracted from the proposed time (2500ms), so the base seek call receives 2700ms
 - Seeking is performend and `GameplayClock.UnderlyingClock.ProcessFrame()` is called.
 - The `totalAppliedOffset` *changes* from -200ms to -100ms since this is the first time `ProcessFrame()` gets called while the track rate is actually 0.5 instead of 1.

This ends up making a difference in `CurrentTime` from the expected one since -200ms was subtracted from it at first but now `totalAppliedOffset` has been corrected to be -100ms instead (so it ends up with 2600ms instead of the proposed 2500ms).

---

With `pauseFreqAdjust` being changed to be 0 by default, `RateAdjustedOffset` will never be touched and tests will pass just fine.

While it may seem awkward that `HardwareCorrectionOffsetClock` only gets processed during seeks from the explaination above, this is only happening because of this edge case (`IsPaused = true && pauseFreqAdjust = 1`). If `IsPaused = false && pauseFreqAdjust = 1`, offsets will be updated per-frame, and if `IsPaused = true && pauseFreqAdjust = 0`, offsets will never be updated neither per-frame nor in seeks.

The only potential issue is when offsets are changed *and* the pause state changed without any prior frames processing the offset change, but I'm not sure that can be solveable anyways (and I might as well be proven wrong on this statement).

Comment (peppy) [reply osu.Game/Screens/Play/MasterGameplayClockContainer.cs:49]: What an ordeal. Thanks for the investigation. Will see what I can do with this knowledge.

Comment (peppy) [reply osu.Game/Screens/Play/MasterGameplayClockContainer.cs:49]: Applying your proposed change breaks other tests, so will need a bit of further thought:

JetBrains Rider-EAP 2022-03-24 at 05 34 15

Comment (peppy) [reply osu.Game/Screens/Play/MasterGameplayClockContainer.cs]: I've fixed the remaining tests and added some test coverage for this scenario. I'm still not 100% on this, but it's something to be considered at a later point. Specifically, I'm not sure how I feel about the gameplay clock potentially being in an incorrect state for the "current" frame because it's being updated as a child of `Player`.

At some point we will want to do an intensive consistency pass on the game loop in general to ensure we aren't inducing any single-frame delays as a result of having components processed too late.

**Comment by peppy** [comment]
[Two](https://github.com/ppy/osu/runs/5671658212?check_suite_focus=true) [of](https://github.com/ppy/osu/runs/5671658249?check_suite_focus=true) the test failures look like they may be related to these changes, but honestly I don't know how. Can't repro locally but I'll leave running in a hope to manage to.

**Comment by peppy** [comment]
Test failures are related to the windows-specific platform offset being applied. Not yet sure on whether this is more or less correct. Maybe @frenzibyte will be able to come to a conclusion quicker than I can.

https://github.com/ppy/osu/runs/5846066167?check_suite_focus=true#step:6:1127

**Comment by frenzibyte** [comment]
Hmm, the test failures raise a very good question on how `MasterGameplayClockContainer` should actually behave when initially stopped in regards to current time and offsets.

On master, since `pauseFreqAdjust = 1` by default, `CreateGameplayClock` would initialise offset clocks and update their `RateAdjustedOffset` as well. Making the first gameplay start time seek in `LoadComplete` work properly.

Meanwhile on this PR, since `pauseFreqAdjust = 0` by default, `CreateGameplayClock` would initialise offset clocks with `RateAdjustedOffset` remaining to be zero. Making the first gameplay start time seek in `LoadComplete` become off by the applied offsets (even though the gameplay clock is not paused at that point).

For a solution, I think that whenever the state of `IsPaused` changes, `pauseFreqAdjust` should update immediately and process the offset clocks to update the `RateAdjustedOffset`, so that the first gameplay start time seek happens properly with `RateAdjustedOffset` being populated:
```diff
diff --git a/osu.Game/Screens/Play/GameplayClockContainer.cs b/osu.Game/Screens/Play/GameplayClockContainer.cs
index 226f5709c7..f2df00d691 100644
--- a/osu.Game/Screens/Play/GameplayClockContainer.cs
+++ b/osu.Game/Screens/Play/GameplayClockContainer.cs
@@ -122,14 +122,14 @@ public void Reset(bool startClock = false, double? gameplayStartTime = null)
             if (gameplayStartTime != null)
                 StartTime = gameplayStartTime;

-            ensureSourceClockSet();
-            Seek(StartTime ?? 0);
-
             // Manually stop the source in order to not affect the IsPaused state.
             AdjustableSource.Stop();

             if (!IsPaused.Value || startClock)
                 Start();
+
+            ensureSourceClockSet();
+            Seek(StartTime ?? 0);
         }

         /// <summary>
diff --git a/osu.Game/Screens/Play/MasterGameplayClockContainer.cs b/osu.Game/Screens/Play/MasterGameplayClockContainer.cs
index f4445d3fa0..2802f721c8 100644
--- a/osu.Game/Screens/Play/MasterGameplayClockContainer.cs
+++ b/osu.Game/Screens/Play/MasterGameplayClockContainer.cs
@@ -131,16 +131,29 @@ private double findEarliestStartTime()
         protected override void OnIsPausedChanged(ValueChangedEvent<bool> isPaused)
         {
             // The source is stopped by a frequency fade first.
-            if (isPaused.NewValue)
+            if (IsLoaded)
             {
-                this.TransformBindableTo(pauseFreqAdjust, 0, 200, Easing.Out).OnComplete(_ =>
+                if (isPaused.NewValue)
                 {
-                    if (IsPaused.Value == isPaused.NewValue)
-                        AdjustableSource.Stop();
-                });
+                    this.TransformBindableTo(pauseFreqAdjust, 0, 200, Easing.Out).OnComplete(_ =>
+                    {
+                        if (IsPaused.Value == isPaused.NewValue)
+                            AdjustableSource.Stop();
+                    });
+                }
+                else
+                    this.TransformBindableTo(pauseFreqAdjust, 1, 200, Easing.In);
             }
             else
-                this.TransformBindableTo(pauseFreqAdjust, 1, 200, Easing.In);
+            {
+                pauseFreqAdjust.Value = isPaused.NewValue ? 1 : 0;
+
+                if (isPaused.NewValue)
+                    AdjustableSource.Stop();
+
+                // process underlying gameplay clocks to update rate-adjusted offsets with the new frequency adjustment.
+                GameplayClock.UnderlyingClock.ProcessFrame();
+            }
         }

         public override void Start()
```

The above diff also reorders `Reset` such that it updates the running state of the clock before performing a seek, for the `RateAdjustedOffset` values to be updated properly for correct seeking.

**Comment by peppy** [comment]
I've applied your proposed fixes with some minor changes. Seems good, and makes sense 👍 .

File: osu.Game/Screens/Play/GameplayClockContainer.cs [hunk 3]
```diff
@@ -106,16 +115,21 @@ public virtual void Seek(double time)
...
         /// </summary>
-        public virtual void Reset()
+        /// <param name="startClock">Whether to start the clock immediately, if not already started.</param>
+        /// <param name="gameplayStartTime">A time to use for future <see cref="Reset"/> calls as the definite start of gameplay.</param>
+        public void Reset(bool startClock = false, double? gameplayStartTime = null)
         {
-            ensureSourceClockSet();
-            Seek(0);
+            if (gameplayStartTime != null)
+                StartTime = gameplayStartTime;
```
Comment (bdach) [inline osu.Game/Screens/Play/GameplayClockContainer.cs:123]: Not sure how to feel about this param and it being used like this, it's a weird edgecasey state mutation in a method that is supposed to be a "reset" method. Is it an option to remove it from `Reset()` and instead explicitly set `StartTime` in call sites that need it?

File: osu.Game/Screens/Play/MasterGameplayClockContainer.cs [hunk 4]
```diff
@@ -46,36 +46,41 @@ public class MasterGameplayClockContainer : GameplayClockContainer
...
         [Resolved]
         private OsuConfigManager config { get; set; }
 
-        public MasterGameplayClockContainer(WorkingBeatmap beatmap, double gameplayStartTime, bool startAtGameplayStart = false)
+        /// <summary>
+        /// Create a new master gameplay clock container.
+        /// </summary>
+        /// <param name="beatmap">The beatmap to be used for time and metadata references.</param>
+        /// <param name="latestGameplayStartTime">The latest time which should be used when introducing gameplay. Will be used when skipping forward.</param>
+        /// <param name="startFromLatestStartTime">Whether to start from the provided latest start time rather than zero.</param>
```
Comment (bdach) [inline osu.Game/Screens/Play/MasterGameplayClockContainer.cs:74]: I don't follow what this `startFromLatestStartTime` toggle is for. When would you want/not want this behaviour? I tried to infer from usages but it looks almost entirely random.

File: osu.Game/Screens/Play/Player.cs [hunk 5]
```diff
@@ -607,30 +607,24 @@ private void performUserRequestedSkip()
         private ScheduledDelegate frameStablePlaybackResetDelegate;
 
         /// <summary>
-        /// Seeks to a specific time in gameplay, bypassing frame stability.
+        /// Specify and seek to a custom start time from which gameplay should be observed.
         /// </summary>
         /// <remarks>
-        /// Intermediate hitobject judgements may not be applied or reverted correctly during this seek.
+        /// This performance a non-frame-stable seek. Intermediate hitobject judgements may not be applied or reverted correctly during this seek.
```
Comment (bdach) [inline osu.Game/Screens/Play/Player.cs:613]:
```suggestion
        /// This performs a non-frame-stable seek. Intermediate hitobject judgements may not be applied or reverted correctly during this seek.
```

**Review by bdach — COMMENTED** [review]
This is a surface level review, I haven't read any of the above, just read source. Nothing seems too bad, but there is one or two things I don't quite get.

File: osu.Game/Screens/Play/MasterGameplayClockContainer.cs (hunk 4 above)
Comment (smoogipoo) [reply osu.Game/Screens/Play/MasterGameplayClockContainer.cs:74]: Only thing I can gather is it looks like making up for `gameplayStartTime` not being a `double?`... Definitely weird and xmldoc doesn't help.

Also not sure why there are 3 ways of providing the start time - via `MasterGameplayClock` ctor, via `StartTime`, and via `Reset()`.

File: osu.Game/Screens/Play/MasterGameplayClockContainer.cs [hunk 6]
```diff
@@ -152,10 +183,10 @@ public override void Seek(double time)
         /// </summary>
         public void Skip()
         {
-            if (GameplayClock.CurrentTime > gameplayStartTime - MINIMUM_SKIP_TIME)
+            if (GameplayClock.CurrentTime > latestGameplayStartTime - MINIMUM_SKIP_TIME)
```
Comment (smoogipoo) [inline osu.Game/Screens/Play/MasterGameplayClockContainer.cs:186]: What's the difference in referring to `latestGameplayStartTime` here instead of `StartTime`?

File: osu.Game/Screens/Play/MasterGameplayClockContainer.cs (hunk 4 above)
Comment (peppy) [reply osu.Game/Screens/Play/MasterGameplayClockContainer.cs:74]: I'll look into the final parameter as I agree it's a bit weird and usages are hard to follow.

> Also not sure why there are 3 ways of providing the start time - via `MasterGameplayClock` ctor, via `StartTime`, and via `Reset()`.

The ctor is setting `latestGameplayStartTime`, which has a very different meaning and purpose. Can you double-check the xmldoc and confirm you can understand from it? The ctor one doesn't change post-construction and is used for cases like skipping forward to gameplay.

File: osu.Game/Screens/Play/MasterGameplayClockContainer.cs (hunk 6 above)
Comment (peppy) [reply osu.Game/Screens/Play/MasterGameplayClockContainer.cs:186]: As mentioned in my last comment, they are completely different variables with different meanings and usages. Can you have a read over the code which uses `latestGameplayStartTime` and potentially suggest a better name for it?

File: osu.Game/Screens/Play/GameplayClockContainer.cs (hunk 3 above)
Comment (peppy) [reply osu.Game/Screens/Play/GameplayClockContainer.cs]: I've made this change, I think it feels better.

**Comment by peppy** [comment]
Apologies for force push, stuff got very mixed up (shouldn't affect review as it was done quickly).

File: osu.Game/Screens/Play/MasterGameplayClockContainer.cs (hunk 4 above)
Comment (peppy) [reply osu.Game/Screens/Play/MasterGameplayClockContainer.cs:74]: How about `skipTargetTime`?

File: osu.Game.Tests/Gameplay/TestSceneMasterGameplayClockContainer.cs [hunk 7]
```diff
@@ -73,26 +73,29 @@ public void TestElapseThenReset()
...
             [Values(0.0, 200.0, -200.0)] double userOffset,
-            [Values(false, true)] bool whileStopped)
+            [Values(false, true)] bool whileStopped,
+            [Values(false, true)] bool setAudioOffsetBeforeConstruction)
         {
             ClockBackedTestWorkingBeatmap working = null;
             GameplayClockContainer gameplayClockContainer = null;
 
+            if (setAudioOffsetBeforeConstruction)
+                AddStep($"preset audio offset to {userOffset}", () => localConfig.SetValue(OsuSetting.AudioOffset, userOffset));
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game.Tests/Gameplay/TestSceneMasterGameplayClockContainer.cs:83]: I thought the use of `LocalStorage` here was enough. I'll double-check.

[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game.Tests/Gameplay/TestSceneMasterGameplayClockContainer.cs:83]: This whole scene is marked `[HeadlessTest]`, so this shouldn't be an issue.

File: osu.Game/Screens/Play/MasterGameplayClockContainer.cs (hunk 4 above)
Comment (peppy) [reply osu.Game/Screens/Play/MasterGameplayClockContainer.cs]: > I don't follow what this `startFromLatestStartTime` toggle is for. When would you want/not want this behaviour? I tried to infer from usages but it looks almost entirely random.

I've removed this and updated usages. Hopefully the local usages should be more understandable due to `StartTime` being set directly.

File: osu.Game.Tests/Gameplay/TestSceneMasterGameplayClockContainer.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (smoogipoo) [reply osu.Game.Tests/Gameplay/TestSceneMasterGameplayClockContainer.cs:83]: I did consider that this is a headless test, but it depends on the ordering of events. Like if nunit ever reorders tests to run an `== true` test before an `== false`, then it's going to be in the wrong state. Rider likes to reorder its tests so...

Guess we can roll with it if you disagree, it's more of a correctness thing.

File: osu.Game/Screens/Play/GameplayClockContainer.cs (hunk 3 above)
Comment (smoogipoo) [reply osu.Game/Screens/Play/GameplayClockContainer.cs]: Is this PR missing a commit? The parameter's still here.

Comment (peppy) [reply osu.Game/Screens/Play/GameplayClockContainer.cs]: You might be looking at an old diff, it's definitely not there for me

File: osu.Game.Tests/Gameplay/TestSceneMasterGameplayClockContainer.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Game.Tests/Gameplay/TestSceneMasterGameplayClockContainer.cs:89]: Fair, I've added a reset before each test run.

**Review by smoogipoo — APPROVED** [review]
Lgtm

File: osu.Game/Screens/Play/MasterGameplayClockContainer.cs (hunk 4 above)
Comment (smoogipoo) [reply osu.Game/Screens/Play/MasterGameplayClockContainer.cs]: New naming and xmldocs are much better.

**Comment by smoogipoo** [comment]
Not sure if the `TestScenePause` failure is something we've seen before...

**Comment by peppy** [comment]
Yeah I'm not sure. Unless anyone has bright ideas I'd probably go forward with merge and see how often it pops up...

**Review by bdach — APPROVED** [review]
I think this cleaned up rather nicely with the latest batch of commits :+1:
